pkg/cloud/services/compute/instance.go

```diff
@@ -280,6 +280,10 @@ func (s *Service) createInstanceImpl(eventObject runtime.Object, openStackCluste
 	})
 	if err != nil {
 		record.Warnf(eventObject, "FailedCreateServer", "Failed to create server %s: %v", createdInstance.Name(), err)
+
+		// we failed to create instance, now with this server = nil, we will let the defer function handle the
+		// defer action to clean up all the ports as it's in infrav1.InstanceStateError status
```

Comment:
How can we make sure that the VM is in InstaneStateError and we do not delete the port of a working VM? There might be other errors, such as overloaded API or connection time out.

We observed some cases where the port of the existing and ready VM was deleted after the API was not reachable, responded with error 500 or a connection timeout / connection refused. I haven't found the exact problem in the code yet - but I think it is somehow related to the port deletion within the `defer`.

Reply:
>>>such as overloaded API or connection time out.
>>>responded with error 500 or a connection timeout / connection refused.

API return 500 indicate API layer reported `internal error` that didn't handle the exception well
but the VM enter ERROR status due to something wrong ,so not combined condition if I understand correctly

the logic here should only handle the `revert` action of create instance, which is , something went wrong and we don't want to keep those resources to be orphan resource ..

Reply:
This isn't the correct place to do this, same as the existing port cleanup, tbh. This really needs to be done in DeleteInstance. Anything else means that any interruption in the controller, for example a restart or some unrelated error, can result in resource leakage.